doc/developer/design/20230528_managed_clusters.md

```diff
@@ -0,0 +1,357 @@
+
+We've always intended to build a declarative API on top of the existing API. The
+original [cluster replica design] from April 2022 sketched this declarative API
+that would allow users to create a cluster with three replicas that
+automatically scaled up and down in response to load:
+
+```sql
+CREATE CLUSTER foo REPLICATION FACTOR 3, MIN SIZE 'small', MAX SIZE 'xlarge'
+```
+
+As described, this API requires "[dynamic cluster scheduling]", which is
+difficult and has many unknowns, so we intentionally chose to not work on this
+in the past year.
+
+The need for improvement has recently become more urgent, as the designers of
+both our [Terraform provider] and our [web console] have considered layering on
+a better cluster management experience in their respective tools. But the right
+place to solve this is in the database itself, so that we don't need to write
+bespoke logic in each tool that is downstream of the SQL API, and so that users
+who use the SQL API directly get the improved experience too.
+
+So, this design document proposes a new course: a new cluster management API
+that solves most of the ergonomic issues with the existing API but does not
+require the full generality of dynamic cluster scheduling.
+
+## Goals
+
+* Users can manage a cluster's size and replication factor without thinking
+  about individual replicas.
+* We can implement the design in weeks, not quarters.
+
+## Non-Goals
+
+* Automatically resizing clusters without downtime.
+* Automatically resizing clusters in response to load.
+* Automatically rebalancing replicas across availability zones.
+
+## Overview
+
+We'll add support for the following declarative cluster management commands to
+Materialize:
+
+```sql
+-- Create a managed cluster with one small replica.
+CREATE CLUSTER foo SIZE 'small';
+
+-- Size up the cluster. IN THE FUTURE this will gracefully
+-- change the size without downtime. AT PRESENT, this will cause
+-- downtime for as long as it takes the new replica(s) to rehydrate.
+--
+-- If you NEED graceful resizing, convert this cluster to an unmanaged
+-- cluster and handle dropping/creating replicas yourself.
+ALTER CLUSTER foo SIZE 'medium';
+
+-- Add two new replicas automatically.
+ALTER CLUSTER foo REPLICATION FACTOR 3;
+
+-- Turn off the cluster for the night.
+ALTER CLUSTER foo REPLICATION FACTOR 0;
+
+-- You can also create a cluster with multiple replicas from the get go.
+CREATE CLUSTER foo2 SIZE 'small', REPLICATION FACTOR 2;
+```
+
+## Detailed description
+
+### Managed clusters
+
+We'll introduce the concept of a **managed** cluster. A managed cluster is one
+with a declared size and replication factor, where Materialize is responsible
+for ensuring the replica set matches the declared size and replication factor.
+The replicas of a managed cluster are visible in the system catalog, but cannot
+be directly modified by users.
+
+Existing clusters, where users control the replica set manually via
```

Comment:
Is there a migration path from unmanaged to managed clusters? Or: do we expect these to converge in the fullness of time?

Reply:
Oh, I caught the answer as I read on!

`ALTER CLUSTER c SET (MANAGED)`